autopush/src/http.rs

```diff
@@ -7,77 +7,87 @@
-        let req_path = req.path().to_string();
+    fn call(&mut self, req: hyper::Request<Body>) -> Self::Future {
+        let mut response = hyper::Response::builder();
+        let req_path = req.uri().path().to_string();
         let path_vec: Vec<&str> = req_path.split('/').collect();
         if path_vec.len() != 3 {
-            response.set_status(StatusCode::NotFound);
-            return Box::new(ok(response));
+            response.status(StatusCode::NOT_FOUND);
+            return Box::new(ok(response.body(Body::empty()).unwrap()));
         }
         let (method_name, uaid) = (path_vec[1], path_vec[2]);
         let uaid = match Uuid::parse_str(uaid) {
             Ok(id) => id,
             Err(_) => {
                 debug!("uri not uuid: {}", req.uri().to_string());
-                response.set_status(StatusCode::BadRequest);
-                return Box::new(ok(response));
+                response.status(StatusCode::BAD_REQUEST);
+                return Box::new(ok(response.body(Body::empty()).unwrap()));
             }
         };
-        let srv = self.0.clone();
+        let clients = Arc::clone(&self.0);
         match (req.method(), method_name, uaid) {
-            (&Method::Put, "push", uaid) => {
+            (&Method::PUT, "push", uaid) => {
                 // Due to consumption of body as a future we must return here
-                let body = req.body().concat2();
+                let body = req.into_body().concat2();
                 return Box::new(body.and_then(move |body| {
                     let s = String::from_utf8(body.to_vec()).unwrap();
                     if let Ok(msg) = serde_json::from_str(&s) {
-                        if srv.notify_client(uaid, msg).is_ok() {
-                            Ok(hyper::Response::new().with_status(StatusCode::Ok))
-                        } else {
-                            Ok(hyper::Response::new()
-                                .with_status(StatusCode::NotFound)
-                                .with_body("Client not available."))
-                        }
+                        Either::A(clients.notify(uaid, msg).then(move |result| {
+                            let body = if result.is_ok() {
+                                response.status(StatusCode::OK);
+                                Body::empty()
+                            } else {
+                                response.status(StatusCode::NOT_FOUND);
+                                Body::from("Client not available.")
+                            };
+                            Ok(response.body(body).unwrap())
+                        }))
                     } else {
-                        Ok(hyper::Response::new()
-                            .with_status(hyper::StatusCode::BadRequest)
-                            .with_body("Unable to decode body payload"))
+                        Either::B(ok(response
+                            .status(hyper::StatusCode::NOT_FOUND)
```

Comment:
```suggestion
                            .status(hyper::StatusCode::BAD_REQUEST)
```

Reply:
heh, weren't these in your updated PR?

reverting.